File: app/src/test/java/de/westnordost/streetcomplete/data/osm/tql/TagFilterExpressionTest.java
package de.westnordost.streetcomplete.data.osm.tql;

import org.junit.Test;

import java.util.Arrays;
import java.util.Collections;

import de.westnordost.osmapi.map.data.Element;

import static org.junit.Assert.*;
import static org.mockito.Mockito.*;

public class TagFilterExpressionTest
{
	// Tests for toOverpassQLString are in FiltersParserTest

	private Element node = createElement(Element.Type.NODE);
	private Element way = createElement(Element.Type.WAY);
	private Element relation = createElement(Element.Type.RELATION);

	@Test public void matchesNodes()
	{
		TagFilterExpression expr = createMatchExpression(ElementsTypeFilter.NODES);

		assertTrue(expr.matches(node));
		assertFalse(expr.matches(way));
		assertFalse(expr.matches(relation));
	}

	@Test public void matchesWays()
	{
		TagFilterExpression expr = createMatchExpression(ElementsTypeFilter.WAYS);

		assertFalse(expr.matches(node));
		assertTrue(expr.matches(way));
		assertFalse(expr.matches(relation));
	}

	@Test public void matchesRelations()
	{
		TagFilterExpression expr = createMatchExpression(ElementsTypeFilter.RELATIONS);

		assertFalse(expr.matches(node));
		assertFalse(expr.matches(way));
		assertTrue(expr.matches(relation));
	}

	@Test public void matchesElements()
	{
		BooleanExpression booleanExpression = mock(BooleanExpression.class);
		when(booleanExpression.matches(any())).thenReturn(true);
		TagFilterExpression expr = new TagFilterExpression(
				Arrays.asList(ElementsTypeFilter.values()),
				booleanExpression);

		assertTrue(expr.matches(node));
		assertTrue(expr.matches(way));
		assertTrue(expr.matches(relation));
	}

	private Element createElement(Element.Type type)
	{
		Element element = mock(Element.class);
		when(element.getType()).thenReturn(type);
		return element;
	}

	private TagFilterExpression createMatchExpression(ElementsTypeFilter elementsTypeFilter)
	{
		BooleanExpression expr = mock(BooleanExpression.class);
		when(expr.matches(any())).thenReturn(true);
		return new TagFilterExpression(Collections.singletonList(elementsTypeFilter), expr);
	}
}

File: app/src/test/java/de/westnordost/streetcomplete/data/osm/tql/TagFilterExpressionTest.kt
package de.westnordost.streetcomplete.data.osm.tql

import org.junit.Test

import de.westnordost.osmapi.map.data.Element
import de.westnordost.streetcomplete.any
import de.westnordost.streetcomplete.mock
import de.westnordost.streetcomplete.on

import org.junit.Assert.*

class TagFilterExpressionTest {
    // Tests for toOverpassQLString are in FiltersParserTest

    private val node = createElement(Element.Type.NODE)
    private val way = createElement(Element.Type.WAY)
    private val relation = createElement(Element.Type.RELATION)

    @Test fun `matches nodes`() {
        val expr = createMatchExpression(ElementsTypeFilter.NODES)

        assertTrue(expr.matches(node))
        assertFalse(expr.matches(way))
        assertFalse(expr.matches(relation))
    }

    @Test fun `matches ways`() {
        val expr = createMatchExpression(ElementsTypeFilter.WAYS)

        assertFalse(expr.matches(node))
        assertTrue(expr.matches(way))
        assertFalse(expr.matches(relation))
    }

    @Test fun `matches relations`() {
        val expr = createMatchExpression(ElementsTypeFilter.RELATIONS)

        assertFalse(expr.matches(node))
        assertFalse(expr.matches(way))
        assertTrue(expr.matches(relation))
    }

    @Test fun `matches nwr`() {
        val expr = createMatchExpression(*ElementsTypeFilter.values())

        assertTrue(expr.matches(node))
        assertTrue(expr.matches(way))
        assertTrue(expr.matches(relation))
    }

    @Test fun `matches filter`() {
        val tagFilter: TagFilter = mock()
        val expr = TagFilterExpression(listOf(ElementsTypeFilter.NODES), Leaf(tagFilter))

        on(tagFilter.matches(any())).thenReturn(true)
        assertTrue(expr.matches(node))
        on(tagFilter.matches(any())).thenReturn(false)
        assertFalse(expr.matches(node))
    }

    private fun createElement(type: Element.Type): Element {
        val element: Element = mock()
        on(element.type).thenReturn(type)
        return element
    }

    private fun createMatchExpression(vararg elementsTypeFilter: ElementsTypeFilter): TagFilterExpression {
        val tagFilter: TagFilter = mock()
        on(tagFilter.matches(any())).thenReturn(true)
        return TagFilterExpression(elementsTypeFilter.asList(), Leaf(tagFilter))
    }
}
